pkg/operator/rules.go

```diff
@@ -69,3 +87,104 @@ func ValidateRule(promRule monitoringv1.PrometheusRuleSpec) []error {
 	_, errs := rulefmt.Parse(content)
 	return errs
 }
+
+// SelectRules selects PrometheusRules for Prometheus or Thanos
+func SelectRules(objectType string, namespaces []string, p PromRuleCommonFields) (map[string]string, error) {
+	promRules := map[string]*monitoringv1.PrometheusRule{}
+	ruleSelector, err := metav1.LabelSelectorAsSelector(p.RuleSelector)
+	if err != nil {
+		return nil, errors.Wrap(err, "convert rule label selector to selector")
+	}
+
+	excludedFromEnforcement := p.ExcludedFromEnforcement
+	// append the deprecated PrometheusRulesExcludedFromEnforce
+	for _, rule := range p.PrometheusRulesExcludedFromEnforce {
+		excludedFromEnforcement = append(excludedFromEnforcement,
+			monitoringv1.ObjectReference{
+				Namespace: rule.RuleNamespace,
+				Group:     monitoring.GroupName,
+				Resource:  monitoringv1.PrometheusRuleName,
+				Name:      rule.RuleName,
+			})
+	}
+	nsLabeler := namespacelabeler.New(
+		p.EnforcedNamespaceLabel,
+		excludedFromEnforcement,
+		true,
+	)
+
+	for _, ns := range namespaces {
+		err := p.RuleInformer.ListAllByNamespace(ns, ruleSelector, func(obj interface{}) {
+			promRule := obj.(*monitoringv1.PrometheusRule).DeepCopy()
+			if err := k8sutil.AddTypeInformationToObject(promRule); err != nil {
+				level.Error(p.Logger).Log("msg", "failed to set rule type information", "namespace", ns, "err", err)
+				return
+			}
+
+			promRules[fmt.Sprintf("%v-%v-%v.yaml", promRule.Namespace, promRule.Name, promRule.UID)] = promRule
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list prometheus rules in namespace %s", ns)
+		}
+	}
+
+	var rejected int
+	rules := make(map[string]string, len(promRules))
+
+	for ruleName, promRule := range promRules {
+		var err error
+		var content string
+		if err := nsLabeler.EnforceNamespaceLabel(promRule); err != nil {
+			continue
+		}
+		if objectType == "prometheus" {
+			// Unset partialResponseStrategy field.
+			for i := range promRule.Spec.Groups {
+				promRule.Spec.Groups[i].PartialResponseStrategy = ""
+			}
+		}
+		content, err = GenerateRulesConfiguration(promRule.Spec, p.Logger)
+		if err != nil {
+			rejected++
+			level.Warn(p.Logger).Log(
+				"msg", "skipping prometheusrule",
+				"error", err.Error(),
+				"prometheusrule", promRule.Name,
+				"namespace", promRule.Namespace,
+				objectType, p.ResourceName,
```

Comment:
we can pass a wrapped logger for PrometheusRuleSelector (e.g. `log.With(c.logger, "prometheus", p.Name, "namespace", p.Namespace)`).